ap/columnar_util.py

```diff
@@ -842,14 +862,53 @@ def all_shifts(self):
 
         return shifts
 
-    def requires(self, func: Callable[[law.Task, dict], dict]) -> None:
+    def update(self, func: Callable[["TaskArrayFunction", Any], None]) -> None:
+        """
+        Decorator to wrap a function *func* that should be registered as :py:attr:`update_func`
+        which is used to update *this* instance dependent on specific task attributes. The function
+        should accept
+
+            - *self* (positional), the instance of the array function itself,
+            - and additional keyword arguments that are set depending on the task that invokes the
+              update.
+
+        In any case it is recommended for the wrapped function to catch all additional keyword
+        arguments. The decorator does not return the wrapped function.
+
+        .. note::
+
+            When the task invoking the requirement is workflow, be aware that both the actual
+            workflow instance as well as branch tasks might call the wrapped function. When the
+            requirements should differ between them, make sure to use the
+            :py:meth:`BaseWorkflow.is_workflow` and :py:meth:`BaseWorkflow.is_branch` methods to
+            distinguish the cases.
+        """
+        self.update_func = func
+
+    def run_update(self, **kwargs) -> None:
+        """
+        Recursively runs the update function of this and all other known instances (via
+        :py:attr:`uses` and :py:attr:`produces`) forwarding *this* instance and all additional
+        *kwargs*.
+        """
+        # run the update of all other known instances
+        for obj in self.uses | self.produces:
```

Comment:
I think one should also run over `self.shifts` here. Right now if other array functions are specified via `shifts` the `update` function is not run on them, which may lead to wrong `shift` resolution. One workaround would be to add these to `uses`, but if one is just interested in the shifts and not the results, passing them via `shifts` would be semantically cleaner. 
See also PR #35 and in particular this commit https://github.com/uhh-cms/analysis_playground/pull/35/commits/607e534fec9c6bcc407afcafc984f25560f349c7 for an example where this becomes relevant.

Reply:
True! I'll address this in the next commit.